1112-i386-txt-Add-Intel-TXT-ACM-module-support.patch

```diff
@@ -0,0 +1,596 @@
++      return NULL;
++    }
++
++  /* Check that table is within module. */
++  if ( user_area_off + sizeof(struct grub_txt_acm_info_table) > hdr->size * 4 )
++    {
++      /* TODO: Is (grub_uint32_t) correct??? */
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM info table size too large: %x > %x"),
++		  user_area_off + (grub_uint32_t)sizeof(struct grub_txt_acm_info_table), hdr->size * 4);
++      return NULL;
++    }
++
++  /* Overflow? */
++  if ( plus_overflow_u32 ((grub_uint32_t)(unsigned long)hdr, user_area_off) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("hdr plus user_area_off overflows"));
++      return NULL;
++    }
++
++    return (struct grub_txt_acm_info_table *)((unsigned long)hdr + user_area_off);
++}
++
++static struct grub_txt_acm_chipset_id_list*
++get_acmod_chipset_list (struct grub_txt_acm_header *hdr)
++{
++  struct grub_txt_acm_info_table *info_table;
++  grub_uint32_t size, id_list_off;
++  struct grub_txt_acm_chipset_id_list *chipset_id_list;
++
++  /* This fn assumes that the ACM has already passed the is_acmod() checks */
++
++  info_table = get_acmod_info_table (hdr);
++  if ( !info_table )
++    return NULL;
++  id_list_off = info_table->chipset_id_list;
++
++  size = hdr->size * 4;
++
++  /* Overflow? */
++  if ( plus_overflow_u32 (id_list_off, sizeof(struct grub_txt_acm_chipset_id)) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("id_list_off plus acm_chipset_id_t size overflows"));
++      return NULL;
++    }
++
++  /* Check that chipset id table is w/in ACM */
++  if ( id_list_off + sizeof(struct grub_txt_acm_chipset_id) > size )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("chipset id list is too big: %x"), id_list_off);
++      return NULL;
++    }
++
++  /* Overflow? */
++  if ( plus_overflow_u32 ((grub_uint32_t)(unsigned long)hdr, id_list_off) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("hdr plus id_list_off overflows"));
++      return NULL;
++    }
++
++  chipset_id_list = (struct grub_txt_acm_chipset_id_list*)
++                    ((unsigned long)hdr + id_list_off);
++
++  /* Overflows? */
++  if ( multiply_overflow_u32 (chipset_id_list->count,
++                              sizeof(struct grub_txt_acm_chipset_id)) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("size of acm_chipset_id_list overflows"));
++      return NULL;
++    }
++
++  if ( plus_overflow_u32 (id_list_off + sizeof(struct grub_txt_acm_chipset_id),
++        chipset_id_list->count * sizeof(struct grub_txt_acm_chipset_id)) )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("size of all entries overflows"));
++      return NULL;
++    }
++
++  /* Check that all entries are w/in ACM */
++  if ( id_list_off + sizeof(struct grub_txt_acm_chipset_id) +
++       chipset_id_list->count * sizeof(struct grub_txt_acm_chipset_id) > size )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM chipset id entries are too big: %x"),
++		  chipset_id_list->count);
++      return NULL;
++    }
```

Comment:
Likewise, `id_list_off + sizeof(struct grub_txt_acm_chipset_id)` should use `sizeof(...id_list)` instead.